File: contract/src/transfer.cpp
#include <eosstrawpoll/contract.hpp>

namespace eosstrawpoll
{

void contract::transfer(
    const eosio::currency::transfer &t,
    const account_name code)
{
    if (t.to != _self || code != N(eosio.token))
    {
        return;
    }

    const auto SYS = S(4, SYS);
    const auto EOS = S(4, EOS);
    const auto TLOS = S(4, TLOS);

    const auto donated_symbol = t.quantity.symbol;

    // only track CORE_SYMBOL transfers
    if (donated_symbol != SYS && donated_symbol != EOS && donated_symbol != TLOS)
    {
        return;
    }

    // only track CORE_SYMBOL transfers
    // if (t.quantity.symbol != CORE_SYMBOL)
    // {
    //     return;
    // }

    const account_name account = t.from;
    const uint64_t donated = t.quantity.amount;

    // add new donation
    new_donations_table.emplace(_self, [&](auto &d) {
        d.id = new_donations_table.available_primary_key();
        d.account = account;
        d.donated = donated;
        d.memo = t.memo;
        d.create_time = now();
    });

    prune_new_donations();

    const auto donation = donation_t{
        .id = 0,
        .account = account,
        .donated = donated,
        .memo = t.memo,
        .create_time = now()};

    // update donors table
    auto donor = donors_table.find(account);
    if (donor == donors_table.end())
    {
        donors_table.emplace(_self, [&](auto &d) {
            d.account = account;
            d.donated = donated;
            d.first_donation = donation;
            d.last_donation = donation;
        });
    }
    else
    {
        donors_table.modify(donor, _self, [&](auto &d) {
            d.donated += donated;
            d.last_donation = donation;
        });
    }
};

void contract::prune_new_donations()
{
    auto num_left = global_config.max_new_donations;
    for (auto it = new_donations_table.rbegin(); it != new_donations_table.rend();)
    {
        if (num_left <= 0)
        {
            it = decltype(it){new_donations_table.erase(std::next(it).base())};
        }
        else
        {
            num_left -= 1;
            ++it;
        }
    }
}

} // namespace eosstrawpoll

File: crates/contract/src/transfer.rs
use crate::types::*;
use crate::constants::*;
use eosio::*;

#[eosio_action]
pub fn transfer(from: AccountName, to: AccountName, quantity: Asset, memo: String) {
    let _self = AccountName::receiver();

    if to != _self {
        return;
    }

    let donated_symbol = quantity.symbol;
    if donated_symbol != SYS && donated_symbol != EOS && donated_symbol != TLOS {
        return;
    }

    let donors = Donor::table(_self, _self);
    let cursor = donors.find(from);
    match cursor {
        Some(cursor) => {
            let mut donor = cursor.get().assert("read");
            donor.donated += quantity.amount;
            cursor.modify(None, &donor).assert("write");
        }
        None => {
            let donor = Donor {
                account: from,
                donated: quantity.amount,
            };
            donors.emplace(_self, &donor).assert("write");
        }
    }

    let donations = Donation::table(_self, _self);
    let id = donations.available_primary_key().assert("error getting next available primary key"); 
    eosio_print!("!!! ", id);
    let donation = Donation {
        id,
        account: from,
        donated: quantity.amount,
        memo,
        create_time: Time::now()
    };
    donations.emplace(_self, &donation).assert("write");
}
